Resolve ILLink warnings in System.Diagnostics.DiagnosticSource

eerhardt opened the pull request:
Contributes to #45623

Addresses all warnings in `System.Diagnostics.DiagnosticSource` with the following plan:

0. Add to the runtime message when a property can't be found that it may have been trimmed away.
1. Mark `DiagnosticSource.Write(string,object)` as `RequiresUnreferencedCode`
2. Suppress the warnings for any .NET libraries that call the `DiagnosticSource.Write()` API, and annotate the .NET types being passed in to preserve their important properties.
    - This was done for HttpClient. ASP.NET and EF will need separate changes when those assemblies are made trim compatible
3. Annotate Activity and its small closure of types (ActivityLink, ActivityEvent, ActivityContext, etc) to ensure none of those properties are trimmed.
4. Suppress trim warnings inside `DiagnosticSourceEventSource` since the public Write method is marked with `RequiresUnreferencedCode`.
5. Logged https://github.com/dotnet/runtime/issues/50454 to add a new Write API that will automatically preserve all the top-level properties on the object (but will still be `RequiresUnreferencedCode` since sub-properties can be selected).

The `DynamicDependency` attributes on `FetcherForProperty` are no longer needed, as the trimmer knows about this Reflection pattern intrinsically and preserves the constructors correctly.

eerhardt committed:
Resolve ILLink warnings in System.Diagnostics.DiagnosticSource (Round 1)

Contributes to #45623

jkotas commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -1382,10 +1379,14 @@ public static PropertyFetch FetcherForProperty(Type? type, string propertyName)
                         else
                         {
-                            PropertyInfo? propertyInfo = typeInfo.GetDeclaredProperty(propertyName);
+                            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:RequiresUnreferencedCode",
```
I do not think that depending on the runtime message is the right approach. This should keep generating warning at build time instead.

eerhardt replied:
> This should keep generating warning at build time instead.

The problem with that is that any apps that use `HttpClient`, any ASP.NET app, or any app that uses EF, will get a warning which the developer can't do anything about. That's a large percentage of .NET applications that will get an unactionable warning.

My understanding of the way this works is that some out of process tool, like ETW, EventPipe, `dotnet-trace`, or VS, can enable the EventSource and start capturing random properties (by using a `FilterAndPayloadSpecs` string) from objects that are logged to the DiagnosticsSource.

ex. https://github.com/dotnet/diagnostics/blob/d562a650488028ec64b66fd032d91945d49404e9/src/Microsoft.Diagnostics.Monitoring.EventPipe/Configuration/HttpRequestSourceConfiguration.cs#L13-L66

The developer at build time isn't able to address this warning, as they don't know what properties are going to try to get logged.

jkotas replied:
Can we fix that by introducing a new generic overload for Write method that allows the linker to see the type and avoid trimming it?

Or can we trim the diagnosic source by default to make it predictably not work?

I think keeping the tracing code around when it is most likely half-broken and you will only find out once you try to use it is worse option than just not making it work by default.

eerhardt replied:
> Can we fix that by introducing a new generic overload for Write method that allows the linker to see the type and avoid trimming it?

That won't work here because the `FilterAndPayloadSpecs` can dot-through multiple properties. The trimmer isn't capable of recursively preserving properties on an object graph - see https://github.com/mono/linker/issues/1087.

> Or can we trim the diagnosic source by default to make it predictably not work?

I made that suggestion offline in an email, but I don't think @noahfalk thought it was a good idea to shut it off completely when it could still possibly work.

> I think keeping the tracing code around when it is most likely half-broken and you will only find out once you try to use it is worse option than just not making it work by default.

Even if it is half-broken, you can still get some information out of it. If it was completely off, when you needed it in production you wouldn't be able to get any information, and would need to rebuild your app to get it, and then still go through all the problems described here.

jkotas replied:
Do we plan to mono/linker#1087 for .NET 6?

>  it was a good idea to shut it off completely when it could still possibly work.

I disagree. Unpredictably half-broken is worse than predictably disabled.

eerhardt replied:
I don’t believe it is going to be implemented for .NET 6

eerhardt replied:
The discussion with @noahfalk compared these diagnostics to debugging, and I think that fits here. Debugging is unpredictably broken for Release builds (can’t get certain variables, line stepping jumps around, breakpoints don’t bind), but we don’t wholesale turn debugging off just because some experiences are degraded.

Your app behavior isn’t affected by this - just your ability to diagnose it.

jkotas replied:
Is the diagnostic source typically only used by humans (ie similar use case as debuggers), or is it often used by automated telemetry systems? My assumption was that it would be often used by automated telemetry systems.

eerhardt replied:
I'm not an expert here, so maybe @noahfalk or @sywhang could answer definitively.

But I know you can use the `dotnet-trace` tool manually to collect traces of your application, which would include this information. It is even doc'd that you can write your own `FilterAndPayloadSpecs`:

https://github.com/dotnet/diagnostics/blob/4a0ee1a66664cb85bfc6268a207b976fa4f5f34d/src/Tools/dotnet-trace/CommandLine/Commands/CollectCommand.cs#L418-L431

I was able to break this in a simple app that sends an HttpRequest every 3 seconds.

```
❯ dotnet-trace collect --process-id XX --providers 'Microsoft-Diagnostics-DiagnosticSource:0x00000003:5:FilterAndPayloadSpecs=\"HttpHandlerDiagnosticListener/System.Net.Http.Request:-Request.RequestUri.Segments\"'
```

When executed against a trimmed app, I got:

![image](https://user-images.githubusercontent.com/8291187/112665179-c6cb9980-8e28-11eb-9680-34c23b40c093.png)

jkotas replied:
I have seen services to revert deployment when telemetry APIs got broken because they are unable to run without automated monitoring systems. The diagnostic APIs are not a nice-to-have feature like a good debugging experience in this scenario.

eerhardt replied:
> I have seen services to revert deployment when telemetry APIs got broken because they are unable to run without automated monitoring systems.

Isn't the suggestion to disable DiagnosticSource in a trimmed app conflicting with this statement?

Honestly, I don't think there is a 100% perfect solution here. We can issue a publish-time warning for all apps that bring in DiagnosticSource (again - any app that uses `HttpClient`, ASP.NET, or EF). But since it is so common, I think developers are just going to ignore it.

jkotas replied:
> Isn't the suggestion to disable DiagnosticSource in a trimmed app conflicting with this statement?

Disabling DiagnosticSource in trimmed apps by default would provide predictable experince. If somebody tries to use DiagnosticSource and trimming together, they will quickly find that these two features do not work well together and probably give up on triming their app. I think that it is better outcome than being unpredictably half-broken.

eerhardt replied:
Do the owners of this library have any thoughts/opinions on what we should do here?

cc @noahfalk @tommcdon @tarekgh @sywhang @josalem

tarekgh replied:
Does the trimming of this part can be handled by a config switch? I mean trim everything by default in the trimmed app and give the option to disable the trimming for this part if any app really wants it.

noahfalk replied:
> > Or can we trim the diagnosic source by default to make it predictably not work?

> I made that suggestion offline in an email, but I don't think @noahfalk thought it was a good idea to shut it off completely when it could still possibly work.

I think Jan's suggestion may be different? I did have a concern around turning off DiagnosticSourceEventSource by default because that serves as the generalized conduit between the app code and the out-of-process tools. It would force every app developer to re-enable it in order to use some of our diagnostic tools. I believe Jan is suggesting to warn/trim on the specific invocations of DiagnosticSource.Write() and that sounds more narrowly targetted. I think when we discussed ActivitySource and DiagnosticSource got lumped together, but if we separate them and make sure we (Microsoft) annotate and suppress warnings for the Activity use-case + Microsoft owned DiagnosticSource use cases that should focus the remaining DiagnosticSource.Write warnings and make them more actionable. I could still easily imagine a bit of iteration where the developers who used DiagnosticSource.Write aren't well aware what a particular diagnostic tool consumes so they may not get the annotations right, then the diagnostic tool breaks, then the author of the tool contacts the library developer and they coordinate updating the annotations.

eerhardt replied:
Note that there are 2 separate issues that the ILLinker is warning about in the `DiagnosticSourceEventSource`. I tried to describe them in the PR description above.

Issue 1 (the issue being addressed here) is for explicit `FilterAndPayloadSpecs` strings. This path can't be made trim compatible without https://github.com/mono/linker/issues/1087 because we can't know what properties are going to be selected by these strings - they can access `Parent.Child.SubChild`. We would need to preserve all properties in the whole graph of the objects being written to `DiagnosticSource`.

Issue 2 - `MakeImplicitTransforms` which just loops over the top level properties of the object being written to `DiagnosticSource`. This can be made trim compatible with a new API `Write<T>(string, T)` where we annotate the `T` type being serialized/written with `[DynamicallyAccessedMembers(PublicProperties)]`. We can mark the existing `Write(string, object)` API as `RequiresUnreferencedCode`, which will cause ILLink warnings at the callsites of the existing API.

noahfalk replied:
Thanks @eerhardt! The MakeImplicitTransform case is primarily intended as an aid for people doing ad-hoc development/exploration of what is available. I can't guarantee nobody uses it their finished tools but that is not purpose it was intended for. I wouldn't go out of our way to make improvements that only benefit the MakeImplicitTransform case, so we can focus on issue 1 and then apply that result across both issues.

I think this might be a proposal that all of us could be happy with:
1. Mark DiagnosticSource.Write(string,object) as RequiresUnreferencedCode (in an ideal world we would narrow this further, but usage of that API is likely low enough that this is good enough)
2. Suppress the warnings for any .NET libraries that call the DiagnosticSource.Write() API, and annotate the .NET types being passed in to preserve their important properties. The canonical examples I am aware of are in Kestrel, HttpClient, and EF. Whenever you are ready we can reach out to VS diagnostics team and AppInsights to make sure we at least have their usage covered. Their tools are probably also a decent approximation for other tools.
3. Annotate Activity and its small closure of types (ActivityLink, ActivityEvent, ActivityContext, etc) to ensure none of those properties are trimmed.
4. Do suppress this linker warning in DiagnosticSourceEventSource as planned. In addition to the error message at runtime we would also be relying on developers having been warned in DiagnosticSource.Write() to figure out what properties they want to commit to making available. The other entrypoint via ActivitySource we handled in (3) by annotating Activity.

I think if we did this the final result would be that most apps would not see a warning, and for the folks who do (people that called DiagnosticSource.Write) it would at least be actionable.

How does that sound?

jkotas replied:
> Suppress the warnings for any .NET libraries that call the DiagnosticSource.Write() API, and annotate the .NET types being passed in to preserve their important properties.

This will be somewhat curbesome since it won't be possible to use the anonymous types to describe the payload anymore. What do you think about introducing generic `Write<T>` overload to make the annotations unnecessary?

noahfalk replied:
No objection to the `Write<T>()` overload but if the goal is to be safe by default we'd still have to emit a warning that the caller needs to affirmatively suppress. `DiagnosticSourceEventSource` allows dotting through multiple levels of hierarchy, [enumerating an IEnumerable and calling ToString()](https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/DiagnosticSourceEventSource.cs,73). Examples in practice are likely rare and maybe non-existent, but in theory `DiagnosticSourceEventSource` could evaluate `argument.Foo.Bar.Baz` where `Foo` was preserved by the `DynamicallyAccessedMembers` attribute on `Write<T>()` but `Bar` and `Baz` were still trimmed causing the scenario to fail.

> This will be somewhat curbesome since it won't be possible to use the anonymous types to describe the payload anymore

Is it possible for the linker to trim properties on anonymous types? I was assuming that if the code executes to create an instance of the anonymous type then the property assignments in that statement would have automatically prevented the properties from being trimmed. Does the linker also trim property writes if there is no statically observable property read?

jkotas replied:
> Is it possible for the linker to trim properties on anonymous types?

Yes.

> the property assignments in that statement would have automatically prevented the properties from being trimmed. 

There is no property assignment in the constructor of anonymous type. The constructor assigns the backing field directly. The property is completely unused as far as the linker can tell.

noahfalk replied:
Aha, I hadn't realized that implementation for anonymous type constructor but makes sense and adds some additional value to the `Write<T>()` API. Thanks : )

vitek-karas replied:
There's also the fact that linker will remove the getter if it only sees the setter being used.

As for the recursive property walk - if there was a way to detect the types in some easy way we could write a custom step for linker for this. The main reason I'm hesitant doing https://github.com/mono/linker/issues/1087 is that pure recursive properties are almost never enough. As mentioned above, even in this case there's still `IEnumerable` and `ToString` on top of pure properties. The other use cases for https://github.com/mono/linker/issues/1087 are similar in that they have some other specific behavior on top of just properties (typically some kind of handling around collections).

My current thinking is that it would be better to cover these uses cases with custom steps for now (assuming there's relatively small number of these) and see if we can come up with a pattern which would work for more than one such custom step. We can then replace the custom step with whatever new annotation we come up with. The ultimate solution is source generators obviously - but I don't know if that's a good solution for this case specifically.

noahfalk replied:
> if there was a way to detect the types in some easy way we could write a custom step for linker for this

`DiagnosticSourceEventSource` is not limited to dotting through properties on the static types either, it can [resolve derived types](https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/DiagnosticSourceEventSource.cs,1296) it finds at evaluation time and then invoke those properties. I expect the pattern is very resistant to static analysis in pathological cases.

I think it would be OK to have `Write()` and `Write<T>()` emit warnings and then give developers writing those calls some guidance. We could do some code searching, I expect we'd find this is an advanced API with fairly limited use. As long as `ActivitySource.StartActivity()` isn't emitting warnings and creating `DiagnosticSourceEventSource` isn't emitting warnings then we've probably mitigated all the common cases.

eerhardt replied:
OK, I believe I have addressed the concerns here with the latest commit. Please take a look and let me know if I've missed something.

I opened #50454 to track adding a new API. That can be done independently of addressing the ILLink warnings in the library.

eerhardt committed:
Update based on the following plan:

1. Mark DiagnosticSource.Write(string,object) as RequiresUnreferencedCode
2. Suppress the warnings for any .NET libraries that call the DiagnosticSource.Write() API, and annotate the .NET types being passed in to preserve their important properties.
    - This was done for HttpClient. ASP.NET and EF will need separate changes when those assemblies are made trim compatible
3. Annotate Activity and its small closure of types (ActivityLink, ActivityEvent, ActivityContext, etc) to ensure none of those properties are trimmed.
4. Suppress trim warnings inside DiagnosticSourceEventSource since the public Write method is marked with RequiresUnreferencedCode.

noahfalk reviewed (approved):
LGTM!

cc @josalem @sywhang @davmason

vitek-karas commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:
```diff
@@ -948,41 +952,55 @@ internal static void CreateActivityListener(DiagnosticSourceEventSource eventSou
                 };
 
-                eventSource._activityListener.ActivityStarted = activity =>
+                eventSource._activityListener.ActivityStarted = activity => OnActivityStarted(eventSource, activity);
+
+                eventSource._activityListener.ActivityStopped = activity => OnActivityStopped(eventSource, activity);
```
Copy paste error? Duplicate lines

noahfalk replied:
Unless my eyes deceive me one is 'Started' and the other is 'Stopped' : )

vitek-karas replied:
You're right - ... I'm... well I guess I need to wake up for real.

eerhardt committed:
Update ref assembly with new attributes.
